enhancements/cluster-logging/cluster-logging-es-pipeline-processing.md

```diff
@@ -0,0 +1,205 @@
+
+In this proposal we assume that the approach B is the one we would start enabling first. But nothing is stopping us from enabling the approach C in the future. Though it is fair to say that approach C will be more complicated to implement because it will require management of custom processing pipelines.
+
+We will be successful when:
+
+* An OCP user (or admin only?)  is able to configure specific logs to be pointed into different (non-default) index in Elasticsearch based on the data contained in the document itself.
+* An OCP user (or admin only?) has enough information and tooling to undertstand and evaluate the impact of creating additional indices. This specifically means new index level metrics will need to be introduced and relevant alerts must be provided as well. Notice that we have recently turned off index level Prometheus metrics (because they were high-cardinality metrics) thus we will need to revisit index level metrics and provide new aggregated non-high cardinality metrics.
+* Any additional non-default indices must be subject to existing retention policy. At the implementation level this means that any new indices have names that comply with data retention implementation OOTB.
+* A multi-tenant data visibility works seamlessly.
+* A roll-over process can manage additional indices seamlessly.
+* An OCP user is able to query data from additional indices.
+* An OCP user is provided enough documentation to be able to understand **when** and **how** to use this feature.
+
+Optional points, not required for inifial implementation:
+
+* An OCP user (or admin only?) can provide custom index mapping for the data. This is specifically needed if the [default Elasticsearch type inference mechanism](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-field-mapping.html) does not lead to acceptable results.
+
+### Non-Goals
+
+* This proposal does not addresses limitations implemented by Elasticsearch to prevent [mapping explosion](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/mapping.html#mapping-limit-settings) (except we need to provide more metrics if needed, see above).
+* This proposal does not enable storing documents with conflicting mappings into common index.
+
+
+## Proposal
+
+When incoming document contains custom data it can happen that it can not be stored into default index because existing index mapping (the index data schema) would conflict with the data in the incoming document.
+
+The main idea in this proposal is to avoid mapping conflicts by pointing incoming document into diffrent index than the default one (how it is decided is discussed later). If that different index hasn't been created yet then it is created on the fly (this is how Elasticsearch works OOTB). In this porposal this index is called an **additional index**.
```

Comment:
are we planning on also implementing these as rollover indices?

Reply:
Absolutely, these additional indices should be subject to roll-over and index aliasing processes as well. I believe this is possible if we nail the index naming schema correctly.

Reply:
@jcantrill would there need to be a lot of change within the rollover logic to address this? or would we just need to make sure these indices are created in a specific way with aliases (as part of the pipeline)?

Reply:
rollover logic relies explicitly upon indices that have a '-write' alias and fit a known pattern (e.g. infra-0000N). if the new indices fit that pattern there is no difference.  If we desire somethign different, it seems possible to me to be able to specify the pattern and re-use the logic.